File: Phase_II/frontend/lib/api/tasks.ts
/**
 * Task API Methods (Server Actions)
 *
 * Type-safe API client methods for task CRUD operations and metrics.
 * All methods are Server Actions that automatically:
 * - Get user_id from Better Auth session
 * - Obtain JWT token via /api/auth/token endpoint
 * - Make authenticated requests to backend API with Authorization header
 *
 * Authentication flow:
 * 1. Session validated via Better Auth cookies (getSession)
 * 2. JWT obtained via /api/auth/token HTTP endpoint
 * 3. Backend validates JWT signature (HS256/RS256)
 */

'use server'

import { headers } from 'next/headers'
import { auth } from '../auth/better-auth'
import {
  mapStatusToErrorCode,
  getUserFriendlyMessage,
  ERROR_CODES
} from './errors'
import type {
  TaskCreate,
  TaskRead,
  TaskUpdate,
  TaskList,
  TaskMetrics,
  ApiResponse,
} from './types'

/**
 * Internal helper to make authenticated API requests
 *
 * Returns structured ApiResponse for type-safe error handling.
 * No longer throws exceptions - all errors returned as ApiError objects.
 */
async function makeAuthenticatedRequest<T>(
  path: string,
  options: RequestInit = {}
): Promise<ApiResponse<T>> {
  try {
    const reqHeaders = await headers();

    // 1. Get Session for the User ID
    const session = await auth.api.getSession({
      headers: reqHeaders
    });

    const userId = session?.user?.id;

    if (!userId) {
      console.error("[API] No user ID in session");
      return {
        success: false,
        error: { code: ERROR_CODES.AUTH_FAILED, message: 'Not authenticated', status: 401 }
      };
    }

    // 2. Get JWT token via Better Auth /api/auth/token endpoint
    //    auth.api.getToken() is undocumented and unreliable server-side.
    //    The documented approach is to call the HTTP endpoint directly.
    let jwt: string | undefined;

    try {
      const authUrl = process.env.BETTER_AUTH_URL || process.env.NEXT_PUBLIC_BETTER_AUTH_URL || 'http://localhost:3000';
      const cookie = reqHeaders.get('cookie') || '';

      const tokenRes = await fetch(`${authUrl}/api/auth/token`, {
        headers: { cookie },
        cache: 'no-store',
      });

      if (tokenRes.ok) {
        const tokenData = await tokenRes.json();
        jwt = tokenData?.token;
      } else {
        console.error('[API] Token endpoint returned:', tokenRes.status);
      }
    } catch (tokenError) {
      console.error("[API] Failed to fetch JWT:", tokenError);
    }

    if (!jwt || !jwt.includes('.')) {
      console.error("[API] No valid JWT token received");
      return {
        success: false,
        error: { code: ERROR_CODES.AUTH_FAILED, message: 'Failed to get JWT token', status: 401 }
      };
    }

    // 3. Construct URL
    const baseURL = process.env.BACKEND_URL || process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
    const url = `${baseURL}/api/${userId}${path}`;

    // 4. Set Headers
    const requestHeaders = new Headers(options.headers || {});
    requestHeaders.set('Content-Type', 'application/json');
    requestHeaders.set('Authorization', `Bearer ${jwt}`);

    const response = await fetch(url, {
      ...options,
      headers: requestHeaders,
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      console.error(`[API] Backend responded: ${response.status}`);
      return {
        success: false,
        error: {
          code: mapStatusToErrorCode(response.status),
          message: getUserFriendlyMessage(response.status, errorData),
          status: response.status
        }
      };
    }

    const data = await response.json();
    return { success: true, data };

  } catch (error) {
    console.error(`[API] Connection Error:`, error);
    return {
      success: false,
      error: { code: ERROR_CODES.BACKEND_UNAVAILABLE, message: 'Server unreachable', status: 503 }
    };
  }
}
/**
 * Create a new task
 * POST /api/{user_id}/tasks
 */
export async function createTask(taskData: TaskCreate): Promise<ApiResponse<TaskRead>> {
  return makeAuthenticatedRequest<TaskRead>('/tasks', {
    method: 'POST',
    body: JSON.stringify(taskData),
  })
}

/**
 * List all tasks for the authenticated user
 * GET /api/{user_id}/tasks
 */
export async function listTasks(): Promise<ApiResponse<TaskList>> {
  return makeAuthenticatedRequest<TaskList>('/tasks', {
    method: 'GET',
  })
}

/**
 * Get a single task by ID
 * GET /api/{user_id}/tasks/{id}
 */
export async function getTask(taskId: string): Promise<ApiResponse<TaskRead>> {
  return makeAuthenticatedRequest<TaskRead>(`/tasks/${taskId}`, {
    method: 'GET',
  })
}

/**
 * Update a task (full replacement)
 * PUT /api/{user_id}/tasks/{id}
 */
export async function updateTask(
  taskId: string,
  updateData: TaskUpdate
): Promise<ApiResponse<TaskRead>> {
  return makeAuthenticatedRequest<TaskRead>(`/tasks/${taskId}`, {
    method: 'PUT',
    body: JSON.stringify(updateData),
  })
}

/**
 * Toggle task completion status
 * PATCH /api/{user_id}/tasks/{id}/complete
 */
export async function toggleTaskComplete(taskId: string): Promise<ApiResponse<TaskRead>> {
  return makeAuthenticatedRequest<TaskRead>(`/tasks/${taskId}/complete`, {
    method: 'PATCH',
  })
}

/**
 * Delete a task
 * DELETE /api/{user_id}/tasks/{id}
 */
export async function deleteTask(taskId: string): Promise<ApiResponse<void>> {
  return makeAuthenticatedRequest<void>(`/tasks/${taskId}`, {
    method: 'DELETE',
  })
}

/**
 * Get aggregated task metrics
 * GET /api/{user_id}/tasks/metrics
 */
export async function getTaskMetrics(): Promise<ApiResponse<TaskMetrics>> {
  return makeAuthenticatedRequest<TaskMetrics>('/tasks/metrics', {
    method: 'GET',
  })
}

File: Phase_V/backend/src/api/v1/tasks.py
"""Task CRUD endpoints with path-based user context.

All endpoints receive user_id from the URL path parameter:
  /api/{user_id}/tasks
  /api/{user_id}/tasks/{id}
  /api/{user_id}/tasks/{id}/complete

Authentication is via Better Auth session cookies.
Authorization ensures user_id in path matches session user_id.
"""

import asyncio
import json
from typing import Annotated, AsyncGenerator, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import StreamingResponse

from src.api.deps import SessionDep
from src.auth import get_current_user_with_path_validation
from src.auth.models import AuthenticatedUser
from src.schemas import TaskCreate, TaskList, TaskMetrics, TaskRead, TaskUpdate
from src.services import task_service

router = APIRouter()

# Type alias for authenticated and authorized user (validates path user_id)
AuthorizedUserDep = Annotated[AuthenticatedUser, Depends(get_current_user_with_path_validation)]


@router.post("", response_model=TaskRead, status_code=status.HTTP_201_CREATED)
async def create_task(
    task_in: TaskCreate,
    session: SessionDep,
    user: AuthorizedUserDep,
) -> TaskRead:
    """Create a new task item.

    The user_id is extracted from the URL path parameter and validated against the session.
    """
    return await task_service.create_task(session, user.user_id, task_in)


@router.get("", response_model=TaskList)
async def list_tasks(
    session: SessionDep,
    user: AuthorizedUserDep,
    status: Optional[str] = Query(default=None, description="pending | completed"),
    priority: Optional[str] = Query(default=None, description="High | Medium | Low"),
    tags: Optional[list[str]] = Query(default=None, description="Filter by tag name(s)"),
    search: Optional[str] = Query(default=None, description="Full-text search in title/description"),
    sort_by: str = Query(default="created_at", description="created_at | priority | title"),
    sort_order: str = Query(default="desc", description="asc | desc"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=50, ge=1, le=200),
) -> TaskList:
    """List tasks for the user with optional filters.

    T059: Supports status, priority, tags, full-text search, sort, and pagination.
    Returns 200 with empty list when no tasks match (never 404).
    """
    return await task_service.list_tasks(
        session,
        user.user_id,
        status=status,
        priority=priority,
        tags=tags,
        search=search,
        sort_by=sort_by,
        sort_order=sort_order,
        page=page,
        page_size=page_size,
    )


@router.get("/metrics", response_model=TaskMetrics)
async def get_metrics(
    session: SessionDep,
    user: AuthorizedUserDep,
) -> TaskMetrics:
    """Get aggregated task statistics for the user.

    Returns metrics including:
    - Total task count
    - Completed task count
    - Pending task count
    - Overdue task count
    - Priority breakdown (high, medium, low)

    All metrics are computed server-side and scoped to the user_id from the URL path.
    """
    return await task_service.get_metrics(session, user.user_id)


@router.get("/{id}", response_model=TaskRead)
async def get_task(
    id: UUID,
    session: SessionDep,
    user: AuthorizedUserDep,
) -> TaskRead:
    """Get a specific task by ID.

    Returns 404 if the task doesn't exist OR if it belongs to a different user.
    This prevents enumeration attacks by not distinguishing between these cases.
    """
    task = await task_service.get_task(session, id, user.user_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )
    return task


@router.put("/{id}", response_model=TaskRead)
async def update_task(
    id: UUID,
    task_in: TaskUpdate,
    session: SessionDep,
    user: AuthorizedUserDep,
) -> TaskRead:
    """Update a task with full replacement (PUT semantics).

    All fields (title, description, is_completed) must be provided.
    Returns 404 if the task doesn't exist or belongs to a different user.
    """
    task = await task_service.update_task(session, id, user.user_id, task_in)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )
    return task


@router.post("/{id}/complete", response_model=TaskRead, status_code=status.HTTP_200_OK)
async def complete_task(
    id: UUID,
    session: SessionDep,
    user: AuthorizedUserDep,
) -> TaskRead:
    """Mark a task as completed (shorthand — no body required).

    T075: Sets status=completed, publishes task.status_changed and task.updated events.
    Idempotent: completing an already-completed task is a no-op.
    Returns 404 if the task doesn't exist or belongs to a different user.
    """
    task = await task_service.complete_task(session, id, user.user_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )
    return task


@router.patch("/{id}/complete", response_model=TaskRead)
async def toggle_task_complete(
    id: UUID,
    session: SessionDep,
    user: AuthorizedUserDep,
) -> TaskRead:
    """Toggle the completion status of a task.

    No request body required - this endpoint toggles the current is_completed value.
    Returns 404 if the task doesn't exist or belongs to a different user.
    """
    task = await task_service.toggle_task_complete(session, id, user.user_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )
    return task


@router.get("/stream", include_in_schema=True)
async def task_stream(
    user: AuthorizedUserDep,
) -> StreamingResponse:
    """SSE stream for real-time task list updates.

    T076: Clients subscribe to receive task.instance_created and task.status_changed events.
    The stream sends a heartbeat every 30 seconds to keep the connection alive.
    Dapr publishes events which are forwarded to connected SSE clients.

    Event format:
      data: {"event_type": "task.status_changed", "task_id": "...", ...}\n\n
    """
    user_id = user.user_id

    async def event_generator() -> AsyncGenerator[str, None]:
        # Send initial connection confirmation
        yield f"data: {json.dumps({'event_type': 'connected', 'user_id': user_id})}\n\n"
        # Send heartbeat every 30 seconds to prevent connection timeout
        while True:
            await asyncio.sleep(30)
            yield f"data: {json.dumps({'event_type': 'heartbeat'})}\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",  # Disable nginx buffering for SSE
        },
    )


@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_task(
    id: UUID,
    session: SessionDep,
    user: AuthorizedUserDep,
) -> None:
    """Delete a task.

    Returns 404 if the task doesn't exist or belongs to a different user.
    On success, returns 204 No Content with no response body.
    """
    deleted = await task_service.delete_task(session, id, user.user_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )
